marshal/admin.go

```diff
@@ -0,0 +1,358 @@
+					"and will not reset offsets.", topic, partID)
+				atomic.StoreInt32(a.claimHealth, 0)
+				return
+			}
+			atomic.AddInt32(a.steps, 1)
+			time.Sleep(<-a.marshaler.cluster.jitters)
+		}
+	}
+}
+
+// claimAndHeartbeat attempts to claim a partition released by a paused consumer.
+// It heartbeats the previous offset.
+func (a *consumerGroupAdmin) claimAndHeartbeat(topic string,
+	partID int, newOffset int64, stopHeartbeat chan struct{}) bool {
+
+	// Get current offsets, which we will try to keep claimHealth.
+	partitionClaim := a.marshaler.GetLastPartitionClaim(topic, partID)
+
+	// Next, try to claim the partition.
+	if !a.marshaler.ClaimPartition(topic, partID) {
+		log.Errorf("[%s:%d] Admin couldn't claim partition to set Kafka offset",
+			topic, partID)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	// Continuously heartbeat the last offsets.
+	a.addClaimAttempt(topic, partID, partitionClaim.CurrentOffset, newOffset)
+	go a.heartbeatLoop(topic, partID, partitionClaim.CurrentOffset, stopHeartbeat)
+	return true
+}
+
+// constructReleaseGroupMessage returns a ReleaseGroup message to write to the Marshal topic.
+func (a *consumerGroupAdmin) constructReleaseGroupMessage() *msgReleaseGroup {
+	now := time.Now()
+	base := &msgBase{
+		Time:       int(now.Unix()),
+		InstanceID: newInstanceID(),
+		ClientID:   a.clientID,
+		GroupID:    a.groupID,
+	}
+	return &msgReleaseGroup{
+		msgBase:       *base,
+		MsgExpireTime: int(now.Add(a.pauseTimeout).Unix()),
+	}
+}
+
+// sendReleaseGroupMessage sends a ReleaseGroup message for a consumer group reading froma given topic.
+func (a *consumerGroupAdmin) sendReleaseGroupMessage(topicName string, partID int) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	topic := a.marshaler.cluster.getPartitionState(a.groupID, topicName, partID)
+
+	for _, partition := range a.releaseGroupPartitions {
+		if int32(topic.claimPartition) == partition {
+			return nil
+		}
+	}
+
+	a.releaseGroupPartitions = append(a.releaseGroupPartitions, int32(topic.claimPartition))
+	rg := a.constructReleaseGroupMessage()
+	_, err := a.marshaler.cluster.producer.Produce(MarshalTopic,
+		int32(topic.claimPartition), &proto.Message{Value: []byte(rg.Encode())})
+	return err
+}
+
+// pauseGroupAndWaitForRelease is called for every partition we'd like to change the offset for.
+// It first sends a ReleaseGroup message to Marshal, then waits for it to be released,
+// then attempts to claim it.
+func (a *consumerGroupAdmin) pauseGroupAndWaitForRelease(topicName string, partID int) bool {
+	if err := a.sendReleaseGroupMessage(topicName, partID); err != nil {
+		log.Errorf("[%s:%d] Admin failed to produce ReleaseMessage group to Kafka: %s",
+			topicName, partID, err)
+		return false
+	}
+
+	// Wait for the paused consumer to release its claim.
+	partitionClaim := a.marshaler.GetPartitionClaim(topicName, partID)
+
+	start := time.Now()
+	for partitionClaim.isClaimed(time.Now().Unix()) {
```

Comment:
This isn't safe because you only have a copy of a claim at a point in time, but someone else could have heartbeated or something and ended up owning the partition. I think you have to do something like,

```
c := time.Ticker(consumerReleaseClaimWaitSleep)
defer c.Release() // I forget the API for ticker...
end := time.After(consumerReleaseClaimWaitTime)

for {
     select {
     case <-c:
         if cl := a.marshaler.GetPartitionClaim(topicName, partID); cl != nil {
            time.Sleep(consumerReleaseClaimWaitSleep) // omg long name
            continue
         }
         return true
     case <-end:
         return false
    }
}
```